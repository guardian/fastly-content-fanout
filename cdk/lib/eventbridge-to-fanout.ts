import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import type { App } from 'aws-cdk-lib';
import { Fn, SecretValue } from 'aws-cdk-lib';
import * as events from 'aws-cdk-lib/aws-events';
import * as targets from 'aws-cdk-lib/aws-events-targets';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as pipes from 'aws-cdk-lib/aws-pipes';
import * as sqs from 'aws-cdk-lib/aws-sqs';
import * as sns from 'aws-cdk-lib/aws-sns';
import { SqsSubscription } from 'aws-cdk-lib/aws-sns-subscriptions';

interface EventbridgeToFanoutStackProps extends GuStackProps {
	snsTopicUpdatesConfig: {
		cfnExportName: string;
		maybeFilterPattern?: object;
		inputTemplatePath: string;
	};
}

export class EventbridgeToFanout extends GuStack {
	constructor(scope: App, id: string, props: EventbridgeToFanoutStackProps) {
		super(scope, id, props);

		const eventBridgeBus = new events.EventBus(this, 'EventBridgeBus', {
			eventBusName: `${this.stack}-eventbridge-bus-${this.stage}`,
		});

		// cdk does not allow us to read the secret from ssm
		const connection = new events.Connection(this, 'Connection', {
			authorization: events.Authorization.apiKey(
				'Fastly-Key',
				SecretValue.secretsManager(`FastlyFanoutApiKey${this.stage}`),
			),
			description: 'Fastly fanout authentication',
			connectionName: `fastly-fanout-connection-${this.stage}`,
		});

		// provisions the destination
		const destination = new events.ApiDestination(this, 'Destination', {
			connection,
			endpoint: `https://api.fastly.com/service/${
				process.env[`FASTLY_FANOUT_SERVICE_ID_${this.stage}`]
			}/publish/`,
			apiDestinationName: `fastly-fanout-api-destination-${this.stage}`,
		});

		new events.Rule(this, 'ApiDestinationRule', {
			eventBus: eventBridgeBus,
			ruleName: `${this.stack}-events-to-fastly-fanout-${this.stage}`,
			// wrapper to feed the destination as a valid target
			targets: [
				new targets.ApiDestination(destination, {
					event: events.RuleTargetInput.fromObject({
						items: [
							{
								channel: events.EventField.fromPath('$.detail.path'),
								formats: {
									// websocket connections
									'ws-message': {
										content: events.EventField.fromPath('$.time'),
									},
									// sse connections
									'http-stream': {
										content: `${events.EventField.fromPath('$.time')}\n`,
									},
								},
							},
						],
					}),
				}),
			],
			eventPattern: {
				region: [this.region],
			},
		});

		const snsTopic = sns.Topic.fromTopicArn(
			this,
			'SNSTopic',
			Fn.importValue(props.snsTopicUpdatesConfig.cfnExportName),
		);

		const sqsQueue = new sqs.Queue(this, 'SqsQueue');

		snsTopic.addSubscription(
			new SqsSubscription(sqsQueue, {
				rawMessageDelivery: true, // prevents message being wrapped in SNS envelope
			}),
		);

		const putEventsOnPipeRole = new iam.Role(
			this,
			'PipeFromSQSToEventBridgeRole',
			{
				inlinePolicies: {
					allowPutEventsFromPipe: new iam.PolicyDocument({
						statements: [
							new iam.PolicyStatement({
								effect: iam.Effect.ALLOW,
								actions: ['events:PutEvents'],
								resources: [eventBridgeBus.eventBusArn],
							}),
							new iam.PolicyStatement({
								effect: iam.Effect.ALLOW,
								actions: [
									'sqs:ReceiveMessage',
									'sqs:DeleteMessage',
									'sqs:GetQueueAttributes',
								],
								resources: [sqsQueue.queueArn],
							}),
						],
					}),
				},
				assumedBy: new iam.ServicePrincipal('pipes.amazonaws.com'),
			},
		);

		// NOTE: 'body' of the SQS message is implicitly parsed and contains the 'PressJob' JSON from the SNS message
		// see https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-pipes-input-transformation.html#input-transform-implicit
		new pipes.CfnPipe(this, 'PipeFromSqsToEventBridge', {
			name: `fronts-updates-pipe-from-sqs-to-eventbridge-${this.stage}`,
			roleArn: putEventsOnPipeRole.roleArn,
			source: sqsQueue.queueArn,
			sourceParameters: props.snsTopicUpdatesConfig.maybeFilterPattern && {
				filterCriteria: {
					filters: [
						{
							pattern: JSON.stringify(
								props.snsTopicUpdatesConfig.maybeFilterPattern,
							),
						},
					],
				},
			},
			target: eventBridgeBus.eventBusArn,
			targetParameters: {
				inputTemplate: JSON.stringify({
					path: props.snsTopicUpdatesConfig.inputTemplatePath,
				}),
			},
		});
	}
}
