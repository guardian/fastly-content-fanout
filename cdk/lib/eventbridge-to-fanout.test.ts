import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { EventbridgeToFanout } from './eventbridge-to-fanout';

describe('The EventBridgeToFanout stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new EventbridgeToFanout(app, 'EventBridgeToFanout', {
			stack: 'stack',
			stage: 'TEST',
			snsTopicUpdatesConfig: {
				cfnExportName: 'test-sns-topic-export-name',
				inputTemplatePath: '<$.foo.bar>',
				maybeFilterPattern: {
					foo: {
						bar: 'baz',
					},
				},
			},
		});
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
