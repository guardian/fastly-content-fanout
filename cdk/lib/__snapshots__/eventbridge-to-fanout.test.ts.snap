// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The EventbridgeToFanout stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [],
    "gu:cdk:version": "TEST",
  },
  "Resources": {
    "ApiDestinationRuleA4096DF5": {
      "Properties": {
        "EventBusName": {
          "Ref": "EventBridgeBus558461B0",
        },
        "EventPattern": {
          "region": [
            {
              "Ref": "AWS::Region",
            },
          ],
        },
        "Name": "playground-events-to-fastly-fanout-TEST",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "DestinationApiDestinationA879FAE5",
                "Arn",
              ],
            },
            "Id": "Target0",
            "InputTransformer": {
              "InputPathsMap": {
                "detail-front-path": "$.detail.front-path",
                "time": "$.time",
              },
              "InputTemplate": "{"items":[{"channel":<detail-front-path>,"formats":{"ws-message":{"content":<time>}}}]}",
            },
            "RoleArn": {
              "Fn::GetAtt": [
                "DestinationEventsRole7DA63556",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "Connection07624BCD": {
      "Properties": {
        "AuthParameters": {
          "ApiKeyAuthParameters": {
            "ApiKeyName": "Fastly-Key",
            "ApiKeyValue": "{{resolve:secretsmanager:FastlyFanoutApiKey:SecretString:::}}",
          },
        },
        "AuthorizationType": "API_KEY",
        "Description": "Fastly fanout authentication",
        "Name": "fastly-fanout-connection-TEST",
      },
      "Type": "AWS::Events::Connection",
    },
    "DestinationApiDestinationA879FAE5": {
      "Properties": {
        "ConnectionArn": {
          "Fn::GetAtt": [
            "Connection07624BCD",
            "Arn",
          ],
        },
        "HttpMethod": "POST",
        "InvocationEndpoint": "https://api.fastly.com/service/undefined/publish/",
        "Name": "fastly-fanout-api-destination-TEST",
      },
      "Type": "AWS::Events::ApiDestination",
    },
    "DestinationEventsRole7DA63556": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/fastly-content-fanout",
          },
          {
            "Key": "Stack",
            "Value": "playground",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DestinationEventsRoleDefaultPolicy798A1E31": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "events:InvokeApiDestination",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "DestinationApiDestinationA879FAE5",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DestinationEventsRoleDefaultPolicy798A1E31",
        "Roles": [
          {
            "Ref": "DestinationEventsRole7DA63556",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EventBridgeBus558461B0": {
      "Properties": {
        "Name": "playground-eventbridge-bus-TEST",
      },
      "Type": "AWS::Events::EventBus",
    },
  },
}
`;
